### PR TITLE
Updates to handling of optionality and defaults

### PR DESCRIPTION
**Pull request by chris-pardy**
Previously `t.optional` wrapped a converter and caused undefined values to "skip" the converter, as a result `t.optional(t.number.default(() => 0))(undefined)` would return `undefined` instead of `0`. This is particularly problematic when using the `t.partial` decorator.
given
```ts
const normal = t.strict({
   name: t.string.default(t.forPath(['fullName']))
});
const partial = t.partial(normal);
```
then the input `{fullName: 'test'}` would produce the output `{name: undefined}`.

Now every converter has an `.optional` property which returns an optional version, `t.optional` just calls this property. The optional converter by default has the same behavior as `t.optional`. However going forward any converter can decide how it wants to handle optionality. The `default`, `pipe`, and `or` compositions apply optionality to all of their component converters. The `forPath` and `sub` converters apply optionality to the converter that they check the value against. In particular this approach means that our default composition using forPath will now actually use default values if given undefined, but still allow undefined values as a result.

### CODE REVIEW

**Comment by RoboCafaz** [comment]
I think I'd prefer `.optional` also be a function instead of a getter--even if they're functionally the same. I think if we're going to go with a functional design we should stay consistent and have it be `.optional()`.

**Comment by chris-pardy** [comment]
> I think I'd prefer `.optional` also be a function instead of a getter--even if they're functionally the same. I think if we're going to go with a functional-style design we should stay consistent and have it be `.optional()`.

It could be either but the `.optional` vs. `.optional()` to me implies that I'm getting a consistent value vs. doing work for instance: `t.number.optional === t.number.optional` feels right while `t.number.optional() !== t.number.optional()` would make sense.

**Comment by ThomasHayOwnUp** [comment]
Edge case: shouldn't `with` allow an arbitrary number of decorators, or are we meant to chain `with`s in that case?

File: src/core.ts [hunk 1]
```diff
@@ -12,33 +12,100 @@ export interface ConverterFunction<Result, Input = unknown> {
...
+   * Applies a decorator function to the converter
+   * @param decorator the decorator function which will be passed this and return the new converter
+   * @returns the new converter after decorating.
+   */
+  with<NextResult, NextInput>(
+    decorator: (converter: TypeConverter<Result, Input>) => TypeConverter<NextResult, NextInput>
+  ): TypeConverter<NextResult, NextInput>;
+}
+
+export const isTypeConverter = <Result, Input>(
```
Comment (ThomasHayOwnUp) [inline src/core.ts:55]: This needs comments

File: src/unions.ts [hunk 2]
```diff
@@ -270,10 +272,23 @@ export const or: Or = (...options: ConverterFunction<any, any>[]) => {
...
+      } else {
+        return [TypeConverterImpl.createTypeConverter(c)];
+      }
+    });
+  }
+
+  with<NextResult, NextInput>(
+    decorator: (converter: TypeConverter<any, any>) => TypeConverter<NextResult, NextInput>
+  ): TypeConverter<any, any> {
+    return new OrTypeConverter(this.flattenedOptions.map((t) => t.with(decorator)));
```
Comment (ThomasHayOwnUp) [inline src/unions.ts:308]: Should this be wrapped in a `try` `catch`?

File: src/decorators.ts [hunk 3]
```diff
@@ -18,3 +18,17 @@ export function optional<Result, Input = unknown>(
   }, `optional ${innerTypeConverter.displayName}`);
   return createConverterFromTypeConverter(optionalConverter);
 }
+
+export function defaultValue<Input>(
+  defaultValue: CreateTypeConverterInput<Input, undefined>
```
Comment (ThomasHayOwnUp) [inline src/decorators.ts:48]: nit: variable with same name as function

**Comment by ThomasHayOwnUp** [comment]
> previously if you did `t.optional(t.number.default(() => 0))` it would allow undefined
> values to pass over the defaulting of the converter. Now you can write
> `t.number.with(t.defaultValue(() => 0)).optional` to get the behavior where if given
> undefined value the defaultValue creator is invoked. In the case where the default value is
> defined using `t.forPath` to source the value from another location this is very useful
> as it allow the value in that other location to determine if the result of the converter is undefined.

Can you provide a test case for the `forPath` scenario you are describing here, apologies if it already exists.

**Comment by matthew-mahony** [comment]
@chris-pardy I'm having problems getting this to work for my situation.  I have a parent object with a number of child properties.  The converter needs to initialize some of the child properties, whether the input parent object is empty or undefined.  This is what I'm trying:

`
const childConverter = t.shape<ChildObject>({
  prop: t.string.with(t.defaultValue(() => 'works'))
});

const parentConverter = t.unknown
  .pipe(
    t.shape<ParentObject>({
      child: childConverter
    }).with(t.defaultValue(() => { child: { prop: 'defaultFromParent' } })).optional
  );

describe('nested converter', () => {
  it('parent undefined calls parent default', () => {
    expect(parentConverter(undefined)).toEqual({ child: { prop: 'defaultFromParent' } });
  });

  it('parent empty calls child default', () => {
    expect(parentConverter({})).toEqual({ child: { prop: 'works' } });
  });
});
`

Neither test is working.  When passing in undefined, the default is not being called on the parent.

When passing in an empty object, it looks like the child converter complains because the 'child' property is undefined.

Let me know if I am using this wrong.

**Comment by chris-pardy** [comment]
> @chris-pardy I'm having problems getting this to work for my situation. I have a parent object with a number of child properties. The converter needs to initialize some of the child properties, whether the input parent object is empty or undefined. This is what I'm trying:
> 
> `
> const childConverter = t.shape({
> prop: t.string.with(t.defaultValue(() => 'works'))
> });
> 
> const parentConverter = t.unknown
> .pipe(
> t.shape({
> child: childConverter
> }).with(t.defaultValue(() => { child: { prop: 'defaultFromParent' } })).optional
> );
> 
> describe('nested converter', () => {
> it('parent undefined calls parent default', () => {
> expect(parentConverter(undefined)).toEqual({ child: { prop: 'defaultFromParent' } });
> });
> 
> it('parent empty calls child default', () => {
> expect(parentConverter({})).toEqual({ child: { prop: 'works' } });
> });
> });
> `
> 
> Neither test is working. When passing in undefined, the default is not being called on the parent.
> 
> When passing in an empty object, it looks like the child converter complains because the 'child' property is undefined.
> 
> Let me know if I am using this wrong.

@matthew-mahony seems right except for some missing parens try this:
```ts
const childConverter = t.shape({
  prop: t.string.with(t.defaultValue(() => 'works'))
});

const parentConverter = t.shape({
  child: childConverter
}).with(t.defaultValue(() => ({ child: { prop: 'defaultFromParent' } }))).optional;

describe('nested converter', () => {
  it('parent undefined calls parent default', () => {
    expect(parentConverter(undefined)).toEqual({ child: { prop: 'defaultFromParent' } });
  });

  it('parent empty calls child default', () => {
    expect(parentConverter({})).toEqual({ child: { prop: 'works' } });
  });
});
```

File: src/decorators.ts (hunk 3 above)
Comment (chris-pardy) [reply src/decorators.ts:48]: good catch

**Comment by chris-pardy** [comment]
> Edge case: shouldn't `with` allow an arbitrary number of decorators, or are we meant to chain `with`s in that case?

typing that is incredibly difficult since each function needs to apply to the output of the previous functions. Similar to pipe I'm imagining a chaining like `.with(defaultValue(...)).with(stringified).with(otherDecoration)`

**Comment by ThomasHayOwnUp** [comment]
> typing that is incredibly difficult since each function needs to apply to the output of the previous functions. Similar to pipe I'm imagining a chaining like `.with(defaultValue(...)).with(stringified).with(otherDecoration)`

I'd recommend adding that example as a unit test somewhere.

**Comment by matthew-mahony** [comment]
@chris-pardy That gave me an error: "At Path $.child, expected {prop: string with default} but was undefined"

I was able to fix it by adding a default for the 'child' property:

`  t.shape<ParentObject>({
      child: childConverter.with(t.defaultValue(() => ( { prop: undefined})))
    }).with(t.defaultValue(() => ({ child: {} }))).optional`

It looks like the key for what I'm trying to do is add a default for every property up the chain.  Which makes sense, it would be hard to add a default property to something that doesn't exist.

I will play with this more and give more feedback tomorrow.

File: src/decorators.ts [hunk 4]
```diff
@@ -1,17 +1,54 @@
-import { Converter, ConverterFunction, createConverter, getConverterName } from './core';
+import { Converter, TypeConverterImpl, CreateTypeConverterInput, TypeConverter } from './core';
+import { createConverterFromTypeConverter } from './create-converter';
 
 /**
+ * @deprecated use convterter.optional instead
```
Comment (matthew-mahony) [inline src/decorators.ts:5]: nit: convterter.optional -> converter.optional

**Comment by matthew-mahony** [comment]
The code looks good.  I don't think we would use these changes though, it requires a bit of rework on our end.  We have been using a workaround using t.compose to initialize what we need, which is simple and working.

File: src/decorators.ts [hunk 5]
```diff
@@ -9,9 +11,27 @@ export function optional<Result, Input = unknown>(
...
 ): Converter<Result | undefined, Input | undefined> {
   return createConverter((input, path, entity) => {
-    if (input === undefined) {
-      return undefined;
+    try {
+      return converter(input!, path, entity);
+    } catch (err) {
+      // input was undefined, output can be undefined.
+      if (input === undefined) {
+        if (err instanceof ConverterError) {
```
Comment (ThomasHayOwnUp) [inline src/decorators.ts:19]: Shouldn't we log or surface these errors in some way?

File: src/__tests__/decorators.test.ts [hunk 6]
```diff
@@ -1,15 +1,52 @@
...
+  it('chains with default correctly', () => {
+    const c = t.optional(t.string.default(() => 'test'));
+    expect(c(undefined)).toBe('test');
+  });
+  it('chains with default using path correctly', () => {
+    const c = t.optional(t.string.default(t.forPath(['test'])));
+    const structure = t.strict({ notTest: c });
+    expect(structure({})).toEqual({ notTest: undefined });
+  });
+  it('chains with or correctly', () => {
```
Comment (ThomasHayOwnUp) [inline src/__tests__/decorators.test.ts:30]: Would it be worthwhile showing this test case working when we don't pass in undefined?

**Review by ThomasHayOwnUp — APPROVED** [review]
LGTM

**Review by matthew-mahony — DISMISSED** [review]
This works on our end.

One question, is there any difference between
`t.unknown.default(() => ({ } )).pipe(converter)`
and
`converter.default(() => ({}))`

The second is cleaner code, but the first makes more sense as far as the order goes (initialize to a default then call the converter).

**Comment by ThomasHayOwnUp** [comment]
@chris-pardy Plans for this ticket?

**Comment by chris-pardy** [comment]
> @chris-pardy Plans for this ticket?

@ThomasHayOwnUp Either merging it today or closing the PR due to unresolvable problems. I'll let you know.

File: src/__tests__/decorators.test.ts (hunk 6 above)
Comment (chris-pardy) [reply src/__tests__/decorators.test.ts:30]: yes

**Comment by chris-pardy** [comment]
> This works on our end.
> 
> One question, is there any difference between
> `t.unknown.default(() => ({ } )).pipe(converter)`
> and
> `converter.default(() => ({}))`
> 
> The second is cleaner code, but the first makes more sense as far as the order goes (initialize to a default then call the converter).

@matthew-mahony no difference because practically `.default` actually goes before the converter it's called on.

File: src/__tests__/decorators.test.ts [hunk 7]
```diff
@@ -1,15 +1,56 @@
...
+  });
+  it('chains with default using path correctly', () => {
+    const c = t.optional(t.string.default(t.forPath(['test'])));
+    const structure = t.strict({ notTest: c });
+    expect(structure({})).toEqual({ notTest: undefined });
+  });
+  it("doesn't allow values in deeper converters to be null.", () => {
+    const c = t.optional(t.forPath(['test'], t.string));
+    expect(() => c({ test: undefined })).toThrow();
+  })
```
Comment (chris-pardy) [inline src/__tests__/decorators.test.ts:33]: @ThomasHayOwnUp does this "feature" seem like a deal breaker for the converter?

Comment (ThomasHayOwnUp) [reply src/__tests__/decorators.test.ts:33]: Thinking about it again yes it is a deal breaker. It's not really optional if it has to be defined.

**Comment by chris-pardy** [comment]
@ThomasHayOwnUp @matthew-mahony I think practically we need to restructure this library (V1) to have optional by default and required as an upgrade path something like: `t.strict({ downPayment: required(t.number) })` while that's annoying it's trivially easy to turn any converter into a required converter (you just trigger an error if it returns an empty value). on the flip side it's really hard to know that a converter rejected because the result would have been empty especially with the `forPath` converter which operates on a different value than the input value.

**Comment by chris-pardy** [comment]
@ThomasHayOwnUp @matthew-mahony back to using the `.optional` property which solves the use case correctly (IMO) @pignatenko this plays into the PR you just opened. I'm guessing you could normalize inputs to undefined, pass them to the optional version of a converter and then normalize outputs to undefined / null?

File: src/create-converter.ts [hunk 8]
```diff
@@ -51,6 +51,7 @@ export function createConverter<Result, Input = unknown>(
     return converter;
   }
   name = name === undefined ? getConverterName(converter) : name;
+  let optionalCache: Converter<Result | undefined, Input | undefined> | null = null;
```
Comment (ThomasHayOwnUp) [inline src/create-converter.ts:71]: Why are we creating a cache? The answer to this question might also make good in-line documentation.

**Comment by chris-pardy** [comment]
Closing in favor of #27

**Pull request closed without merging**